### ISSUE TITLE
Memory cache lock contention

### ISSUE THREAD

**Issue by ynkm169** [apollographql/apollo-kotlin#4605]
### Use case

Hello, 
`private val lock = CacheLock()`

MemoryCache uses a lock that can be problematic. With Java concurrent data structures, lock stripping/lock splitting would make data structures such as ConcurrentHashMap very efficient. Android LRU cache internally hopefully is pretty efficient as well. 

However, with a synchronized lock like this, effectively the whole LRU cache is locked up allowing only 1 reader/writer at once.

Did some systracing. Launched 50 duplicate API calls, result shows there are around 60 RxCachedThreadS and 60 DefaultDispatch threads all contending for 1 monitor lock. 
<img width="1123" alt="Screen Shot 2023-01-04 at 6 15 10 PM" src="https://user-images.githubusercontent.com/1810298/210673974-c7f5b01a-5328-44a2-a0d4-8441b4048239.png">

### Describe the solution you'd like

Question
Is it possible to remove this lock all together and have some other work around? Possibly even sacrificing consistency for performance?

**Comment by BoD** [comment]
Thank you for providing this, this is interesting data, and will dig into it.

Just a quick reaction to 
> it is not uncommon for dev to call an API to retrieve a list of items, and use flatmap to make a seperate API call for each item in the list

Not sure if you are using GraphQL already (if yes you can disregard my comment), but one of its advantage when compared to REST is the ability to avoid this, by having the fields you need at your disposal to be used in one query. Of course the ability to do it will depend on the server's schema, but in many cases you should be able to avoid some round trips.

**Comment by ynkm169** [comment]
> Not sure if you are using GraphQL already
Yes, to provide more context, REST and GraphQL will co-exist for quite a while for us. Migration will take a lot of time.

**Comment by ynkm169** [comment]
I wasn't sure if it is possible to build a LRU cache that's more performant than a synchronized LinkedHashMap. I Looks like [Guava cache](https://github.com/google/guava/blob/67350a341c5cfd5e9bebf86ba87687ab7ef602b2/guava/src/com/google/common/cache/LocalCache.java) is potentially more performant (hard to tell since Guava cache maybe better used for server than low power mobile devices. 

> The Least Recently Used page replacement algorithm was chosen due to its simplicity.
> The basic strategy is to subdivide the table among Segments, each of which itself is a
>    * concurrently readable hash table. The map supports non-blocking reads and concurrent writes
>    * across different segments.

**Comment by BoD** [comment]
A few additional notes:
- About your last comment about LRU, this is interesting. My hunch is that optimizing the LRU cache itself won't have a big impact (I could be wrong!) but that the locking situation showed with your traces could be improved
- The need for the synchronized block in `MemoryCache.loadRecord()` is that, although it is already called under a read lock (from `DefaultApolloStore.readOperation`), the code actually alters the inner state, and we need to make sure that only one thread is doing that at a time. As you noted above, updating the LRU cache state itself should be a very fast operation however. 
 <br>But inside the same synchronized block the code also reads the data of the next cache. This is usually the disk cache and does IO, so not a fast operation. Locking this part is also unnecessary since we're already under the read lock. We could improve this by reading the disk cache outside of the synchronized block, and only do the memory changes under  synchronized block. This would allow multiple threads to read from the disk cache at the same time, as intended, probably a good improvement to make.

**Comment by ynkm169** [comment]
> But inside the same synchronized block the code also reads the data of the next cache. This is usually the disk cache and does IO, so not a fast operation.
Great idea! What's the class name of the disk cache? If you are referring to http cache, we don't use it.

**Comment by BoD** [comment]
No sorry I was referring to the [SQLite cache](https://www.apollographql.com/docs/kotlin/caching/normalized-cache/#sqlite-cache) ([`SqlNormalizedCache`](https://github.com/apollographql/apollo-kotlin/blob/05f2a3295517fb7a64c2259c38b453a37f3c4d19/libraries/apollo-normalized-cache-sqlite/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/sql/SqlNormalizedCache.kt#L13)). Maybe you don't need it either actually - it's only useful if you need a persistent cache in your app of course.

**Comment by ynkm169** [comment]
We don’t use sql cache at least for now (may do in future to persist user data such as making a post and save). So that lock contention is caused by something else.

Sent from my iPhone

On Jan 5, 2023, at 1:55 PM, Benoit Lubek ***@***.***> wrote:

﻿

No sorry I was referring to the SQLite cache<https://www.apollographql.com/docs/kotlin/caching/normalized-cache/#sqlite-cache> (SqlNormalizedCache<https://github.com/apollographql/apollo-kotlin/blob/05f2a3295517fb7a64c2259c38b453a37f3c4d19/libraries/apollo-normalized-cache-sqlite/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/sql/SqlNormalizedCache.kt#L13>). Maybe you don't need it either actually - it's only useful if you need a persistent cache in your app of course.

**Comment by BoD** [comment]
Thanks for clarifying. It's difficult to pinpoint the source of the contention. FWIW, what I observe in a similar test - albeit without Rx - looks a bit different.

I'm wondering if it would make sense to get rid of a high level lock on `ApolloStore` and let the `NormalizedCache`s handle their own locking instead.

Overall, we think the normalized cache could benefit from a revamp, including performance improvements (in this area but also in terms of storage size, access time, etc.), as well as new features (see #2331) - but this will be a major task.

**Comment by ynkm169** [comment]
> I'm wondering if it would make sense to get rid of a high level lock on ApolloStore and let the NormalizedCaches handle their own locking instead.
I am not familiar with internals of everything the lock currently guards.

Regarding the our own repository layer implementation for comparison with Apollo Store, it does not hold a lock under
1. It basically concats 3 streams of data and optionally take the first that comes back. The result observable can be subscribed on IO thread. We basically do not hold any lock anywhere. Android LRU cache implementation is thread safe.
A: in memory data source with [Android LRU cache implementation](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/util/LruCache.java).
B: DB data source (very rare)
C: remote REST data source.

**Comment by ynkm169** [comment]
Hey, just sharing some ideas I found for improving performance of readers/writers
JakeW's [DiskLruCache](https://github.com/JakeWharton/DiskLruCache) used by Okhttp and Bumptech/Glide image loading.
- Multiple readers can read the cache even if there is a writer that is currently writing. Writer write to a temporary file, and swap with the main file when it's done. It sacrifice consistency for performance.
- Does not allow multiple writers to the same file.

**Comment by ashare80** [comment]
Running into performance issues with this with large queries.

Found that there was some extra logic inside of the locks that could be moved out: https://github.com/apollographql/apollo-kotlin/pull/5101

This was minor performance boost when testing these modifications on our data. 

It appears on a query cache read I am trying to boost performance, a lot of the time is spent in JSON serialization logic trying to build out the records.

<img width="852" alt="Screenshot 2023-07-18 at 1 50 58 AM" src="https://github.com/apollographql/apollo-kotlin/assets/5692739/2c2d7cc2-4fc4-4c1a-be27-ea1fee38f257">

**Comment by martinbonnin** [comment]
Thank you so much @ashare80 💙 . The investigation and PR are awesome! Is there any chance you could contribute your data? We could include that in benchmarks and that'd be a good data point for future optimisation. No worries at all if not, we can always forge tests but real life data always has more flavour :)

**Comment by ashare80** [comment]
@martinbonnin We might be able to work something out for that. What would be the best way to share this? Would a sample response be sufficient?

**Comment by martinbonnin** [comment]
Nice! If you can share the schema + query too, it'd be ideal. If not we can craft something out but it's a bit painful + might not be able to infer the polymorphism if any.
